pkg/server/events.go

```diff
@@ -236,10 +278,78 @@ func handleSandboxExit(e *eventtypes.TaskExit, sb sandboxstore.Sandbox) {
 		return status, nil
 	})
 	if err != nil {
-		logrus.WithError(err).Errorf("Failed to update sandbox %q state", e.ContainerID)
-		// TODO(random-liu): [P0] Enqueue the event and retry.
-		return
+		return fmt.Errorf("failed to update sandbox state: %v", err)
 	}
 	// Using channel to propagate the information of sandbox stop
 	sb.Stop()
+	return nil
+}
+
+func (b *backOff) getExpiredContainers() []string {
+	var containers []string
+	now := time.Now()
+	for c, v := range b.queues {
+		if v.time.isExpire(now) {
+			containers = append(containers, c)
+		}
+	}
+	return containers
+}
+
+func (b *backOff) isInBackOff(evt interface{}) (string, bool) {
+	key := ""
+	switch evt.(type) {
+	case *eventtypes.TaskExit:
+		key = evt.(*eventtypes.TaskExit).ContainerID
+	case *eventtypes.TaskOOM:
+		key = evt.(*eventtypes.TaskOOM).ContainerID
+	default:
+		return key, false
+	}
+	return key, len(b.queues[key].events) != 0
+}
+
+// enBackOff start to backOff and put event to the tail of queue
+func (b *backOff) enBackOff(key string, evt interface{}) {
+	queue := b.queues[key]
+	queue.events = append(queue.events, evt)
+	if len(queue.events) == 1 {
+		queue.time = queue.time.newTime()
+	}
+	b.queues[key] = queue
+}
+
+// enBackOff get out the whole queue
+func (b *backOff) deBackOff(key string) backOffQueue {
+	queue := b.queues[key]
+	delete(b.queues, key)
+	return queue
+}
+
+// enBackOff start to backOff again and put [nth:] events to the queue
+func (b *backOff) reBackOff(key string, queue backOffQueue, n int) {
+	b.queues[key] = backOffQueue{
+		events: queue.events[n:],
+		time:   queue.time.newTime(),
+	}
+}
+
+func (b *backOff) stopBackOff() {
+	b.ticker.Stop()
+}
+
+func (t backOffTime) isExpire(now time.Time) bool {
+	return !now.Before(t.expireTime)
+}
+
+func (t backOffTime) newTime() backOffTime {
```

Comment:
I think this function is a little bit unnecessary complex... Simple and straightforward code is preferred. See comments above.